crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -46,10 +46,23 @@ impl<'db> UnionBuilder<'db> {
     pub(crate) fn add(mut self, ty: Type<'db>) -> Self {
         match ty {
             Type::Union(union) => {
-                self.elements.extend(union.elements(self.db));
+                for element in union.elements(self.db) {
+                    self = self.add(*element);
+                }
             }
             Type::Never => {}
             _ => {
+                let mut remove = vec![];
+                for element in &self.elements {
+                    if ty.is_subtype_of(self.db, *element) {
+                        return self;
+                    } else if element.is_subtype_of(self.db, ty) {
+                        remove.push(*element);
+                    }
+                }
+                for element in remove {
+                    self.elements.remove(&element);
+                }
```

Comment:
This makes `add` `O(n^2)` where `n` is the number of removed elements. It's probably still a very low number. It might be worth to store the removed elements in an HashSet and then using `self.elements.retain`

Reply:
I considered suggesting a HashSet for `remove`, but then I realised that it's not necessary to use a HashSet to guarantee that all items in `remove` are unique. Because of the way it's constructed, we know that all items in `remove` are also members of `self.elements`, and `self.elements` is a `HashSet`

Reply:
It's less about the uniquness but more about using a more efficient removal method that avoids copying the array elements `n` times where `n` is the number of removed elements. 

We can probably do something more clever once `self.elements` is a `Vec` because we know that the elements in `remove` and `self.elements` are in the same order. But it may also just not be worth it (hashing isn't free either). 

The cheapest would be to use `swap_remove` but that changes the order of the elements. Not sure if we're fine with that.

Reply:
I don't think a high cardinality for `remove` will be common, but it certainly could happen (e.g. if you have many subtypes of the same base type all in a union, and then you add the common base type), so it's worth considering. Good comments!

I've replaced the `FxOrderSet` with a `Vec`, and I'm now doing a paired iteration to build up a new vec in a single pass with only the elements we want to keep (and using `with_capacity` to avoid ever needing to resize it). Will push a PR in a bit.